### PR TITLE
[GENX] Add SubGroupShuffle operation

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Add `genx.sub_group_shuffle` operation to the GENX dialect.

### CODE REVIEW

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 1]
```diff
@@ -144,4 +144,22 @@ def GENX_BarrierOp : GENX_Op<"barrier"> {
   let assemblyFormat = "attr-dict";
 }
 
+//===----------------------------------------------------------------------===//
+// Subgroup shuffle xor
+
+def GENX_SubGroupShuffleXorOp : GENX_Op<"sub_group_shuffle_xor"> {
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: Can you design a generic sub-group shuffle operation in the GENX dialect (similar to NVVM approach) and handle the Xor, up, down operations via an enum. Also make the GENX operation accept different data types for the input args  and the return value. That way the GENX dialect will have the ability to model all the shuffle operations supported by the HW (or at least mirror the API exposed by the SPIRV translator) and provide functionality that is close to the functionality exposed by the NVVM dialect.

File: mlir/include/mlir/Dialect/LLVMIR/CMakeLists.txt [hunk 2]
```diff
@@ -66,4 +66,8 @@ add_mlir_dialect(GENXOps genx)
 add_mlir_doc(GENXOps GENXDialect Dialects/ -gen-dialect-doc -dialect=genx)
 set(LLVM_TARGET_DEFINITIONS GENXOps.td)
 mlir_tablegen(GENXConversions.inc -gen-llvmir-conversions)
+mlir_tablegen(GENXOpsEnums.h.inc -gen-enum-decls)
+mlir_tablegen(GENXOpsEnums.cpp.inc -gen-enum-defs)
+mlir_tablegen(GENXOpsAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=genx)
+mlir_tablegen(GENXOpsAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=genx)
 add_public_tablegen_target(MLIRGENXConversionsIncGen)
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/CMakeLists.txt]: Add newline

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 3]
```diff
@@ -144,4 +146,31 @@ def GENX_BarrierOp : GENX_Op<"barrier"> {
...
+  [ShflKindXor, ShflKindUp, ShflKindDown]> {
+  let genSpecializedAttr = 0;
+  let cppNamespace = "::mlir::GENX";
+}
+def ShflKindAttr : EnumAttr<GENX_Dialect, ShflKind, "shfl_kind">;
+
+def GENX_SubGroupShuffleOp : GENX_Op<"sub_group_shuffle", [
+      TypesMatchWith<"result and value have the same type",
+                     "res", "value", "$_self">]>,
+  Results<(outs LLVM_Type:$res)>,
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: LLVM_Type can be a struct, a ptr type, etc... Can we be more restrictive and only allow integer and float types ?

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 4]
```diff
@@ -144,4 +146,31 @@ def GENX_BarrierOp : GENX_Op<"barrier"> {
...
+  Arguments<(ins LLVM_Type:$value,
+                 I32:$mask,
+                 ShflKindAttr:$kind)> {
+  string llvmBuilder = [{
+    createSubGroupShuffle(builder, $value, $mask, $kind);
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: $ret = ...

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 3 above)
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: Add baseDescription and summary

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp [hunk 5]
```diff
@@ -40,6 +40,44 @@ static llvm::Value *createDeviceFunctionCall(llvm::IRBuilderBase &builder,
...
+    break;
+  case GENX::ShflKind::UP:
+    fnName = "_Z20sub_group_shuffle_up";
+    break;
+  case GENX::ShflKind::DOWN:
+    fnName = "_Z22sub_group_shuffle_down";
+    break;
+  };
+  llvm::Type *ty = value->getType();
+  if (ty->isHalfTy())
```
Comment (etiotto) [inline mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:63]: Use a TypeSwitch

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp [hunk 6]
```diff
@@ -40,6 +40,44 @@ static llvm::Value *createDeviceFunctionCall(llvm::IRBuilderBase &builder,
   return builder.CreateCall(fn, args);
 }
 
+static llvm::Value *createSubGroupShuffle(llvm::IRBuilderBase &builder,
+                                          llvm::Value *value, llvm::Value *mask,
+                                          GENX::ShflKind kind) {
+  std::string fnName = "";
```
Comment (etiotto) [inline mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:47]: assert that mask as type that is a i32. 
assert that value has type that is an integer or floating point type ?

Comment (whitneywhtsang) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:47]: no need tablegen should have already did that for you.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp (hunk 5 above)
Comment (whitneywhtsang) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:63]: How strongly you want to use TypeSwitch? I also wanted to use TypeSwitch in the beginning, and find that the current way is easier.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp (hunk 6 above)
Comment (etiotto) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:47]: Yes I know it did it for the operator but this is a function that might get called somewhere else so adding an assertion would help catch potential mismatches in future possible contexts.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp (hunk 5 above)
Comment (etiotto) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:63]: Ok I see. I am fine with the current code then.

File: mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp (hunk 6 above)
Comment (whitneywhtsang) [reply mlir/lib/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.cpp:47]: added assert for mask's type. Already have unreachable for value's type.

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 7]
```diff
@@ -170,11 +170,15 @@ def GENX_SubGroupShuffleOp : GENX_Op<"sub_group_shuffle", [
...
-    The `genx.sub_group_shuffle` operation returns,
-    - $kind == XOR, the $value of the invocation identified by the current invocation’s id within the subgroup xor’ed with $mask.
-    - $kind == UP, the $value of the invocation identified by the current invocation’s id within the subgroup - $mask. 
-    - $kind == DOWN, the Value of the invocation identified by the current invocation’s id within the subgroup + $mask.
-    - $kind == IDX, the $value of the invocation identified by the id $mask.
+    The `genx.sub_group_shuffle` operation is invoked by different work items
+    with different values, given by $value. Different work items have different
+    subgroup local IDs. The shuffle kind, $kind, is given to determine how to
+    calculate the associated subgroup local ID.
+    It returns the associated $value for the work item with subgroup local ID equals:
```
Comment (etiotto) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td]: equals -> equal to

**Review by etiotto — APPROVED** [review]

**Pull request merged**